### ISSUE TITLE
protobuf: upstream_bind_config is a core.v3.BindConfig instead of an UpsteamBindConfig

### ISSUE THREAD

**Issue by davidkornel** [envoyproxy/envoy#15516]
*Description*:
I'm not sure if it's just "misspelt" or it has to be like this, but cluster.upstream_bind_config is a [core.v3.BindConfig](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/cluster/v3/cluster.proto#L885). I think it should be a [UpstreamBindConfig](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/cluster/v3/cluster.proto#L1073). This way UpstreameBindConfig message is totally unused. And using the current state does nothing.

*Additional questions*:
How exactly should [this](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#config-cluster-v3-upstreambindconfig) field work?
With this upsteam_bind_config field should I be able to tell envoy which port_value should it use when sending datagrams to the upstream host? I mean we want to control the port value from our worker rtpengine gets its datagrams. Instead of letting envoy open a socket from a random port, we want to tell which specific port to use when connecting to an upstream host. Is it possible?

**Comment by lambdai** [comment]
Good catch. I think we should deprecate `UpstreamBindConfig`  though it's more accurate. Cc @htuch

**Comment by lambdai** [comment]
I use BindConfig to bind source ip. 
BindConfig::SockAddress contains port (source port) as you need
`uint32 port_value = 3 [(validate.rules).uint32 = {lte: 65535}];

**Comment by htuch** [comment]
Yes, this should be marked as deprecated at message level and comments added to that effect. This is pretty confusing, I agree.

**Comment by davidkornel** [comment]
@lambdai UpstreamBindConfig works with HTTP, but not with UDP.
I tried using it with UDP, but it looks like it doesn't work with it. 

I extended my cluster's config with the following code :

upstream_bind_config:
      source_address:
        address: 127.0.0.2
        port_value: 9999

After that I setup a socat sender to my envoy listener `socat -d -d - udp:127.0.0.1:10020` and a socat receiver as the upstream host `socat -d -d udp-l:19000 -`

Results from the receiver: 
`2021/03/17 14:10:30 socat[28284] N accepting UDP connection from AF=2 127.0.0.1:51755`
It should be `...from 127.0.0.2:9999`. `51755` is chosen by envoy randomly I guess.

If I try the same setup with a simple HTTP router setup it work as I expected:
`socat -d -d tcp-l:19000 -
2021/03/18 13:12:02 socat[10238] N listening on AF=2 0.0.0.0:19000
2021/03/18 13:12:26 socat[10238] N accepting connection from AF=2 127.0.0.2:9999 on AF=2 127.0.0.1:19000`

Maybe this functionality is not implemented in the UDP pipeline? Not sure, but definitely interesting.
WDYT?

**Comment by htuch** [comment]
I think there is confusion here. There is the `UpstreamBindConfig` message, which nothing references, and should be deprecated, and the `upstream_bind_config` field, which actually has message type `BindConfig`. Re: UDP support, tagging @danzh1989 @alyssawilk

**Comment by davidkornel** [comment]
> I think there is confusion here. There is the `UpstreamBindConfig` message, which nothing references, and should be deprecated, and the `upstream_bind_config` field, which actually has message type `BindConfig`. Re: UDP support, tagging @danzh1989 @alyssawilk

Yeah sorry, you are right. I meant upstream_bind_config instead of UpstreamBindConfig. I understand that it should be deprecated. Thanks for tagging them.

**Comment by mattklein123** [comment]
> Maybe this functionality is not implemented in the UDP pipeline? Not sure, but definitely interesting.

Can you open a separate issue on this? I can fix this. I assume your full config is using udp_proxy? The problem with bind config with a single port is it breaks session tracking, so it needs some thinking. Let's discuss in a different issue.
